File: adversarial-harness/python/policy_engine/injection_preprocess.py
"""Injection text preprocessing — mirrors src/utils/injection-preprocess.ts."""

from __future__ import annotations

import re
import unicodedata

from .confusables import fold_homoglyphs

INVISIBLE_TO_SPACE_RE = re.compile(
    r"[\u200B-\u200F\uFEFF\u00AD\u2060-\u2064\u061C\u180E\u034F\u17B4\u17B5\u202A-\u202E\u2800\uFE00-\uFE0F]"
)
UNICODE_SPACE_RE = re.compile(r"[\u00A0\u1680\u2000-\u200A\u202F\u205F\u3000\u3164]")
ANSI_ESCAPE_RE = re.compile(r"\x1b\[[0-9;]*[A-Za-z]|\x1b\].*?\x07")
EMOJI_RE = re.compile(
    r"[\U0001F300-\U0001FAFF\U00002600-\U000027BF\U00002300-\U000023FF]"
)
LEET_MAP = str.maketrans("013457@$", "oieastas")

EXTRA_HOMOGLYPHS: dict[int, str] = {
    0x0438: "i",
    0x0433: "g",
    0x043D: "n",
    0x0442: "t",
    0x0432: "b",
    0x043B: "l",
    0x043C: "m",
    0x0434: "d",
    0x0455: "s",
    0x04CF: "l",
}


def rot13(text: str) -> str:
    out: list[str] = []
    for ch in text:
        if "a" <= ch <= "z":
            out.append(chr((ord(ch) - 97 + 13) % 26 + 97))
        elif "A" <= ch <= "Z":
            out.append(chr((ord(ch) - 65 + 13) % 26 + 65))
        else:
            out.append(ch)
    return "".join(out)


def deleetspeak(text: str) -> str:
    return text.translate(LEET_MAP)


def strip_combining_marks(text: str) -> str:
    decomposed = unicodedata.normalize("NFD", text)
    stripped = "".join(c for c in decomposed if unicodedata.category(c) != "Mn")
    return unicodedata.normalize("NFC", stripped)


def fold_extended_homoglyphs(text: str) -> str:
    out = fold_homoglyphs(text)
    result: list[str] = []
    for ch in out:
        code = ord(ch)
        if 0x1D400 <= code <= 0x1D7FF:
            offset = code - 0x1D400
            latin = offset % 26
            is_upper = (offset // 26) % 2 == 1
            result.append(chr((65 if is_upper else 97) + latin))
            continue
        result.append(EXTRA_HOMOGLYPHS.get(code, ch))
    return "".join(result)


def collapse_control_whitespace(text: str) -> str:
    text = INVISIBLE_TO_SPACE_RE.sub(" ", text)
    text = UNICODE_SPACE_RE.sub(" ", text)
    text = re.sub(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]", " ", text)
    text = re.sub(r"[\t\r\n\f\v]+", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def preprocess_for_injection_match(text: str, unicode_strict: bool = True) -> str:
    current = INVISIBLE_TO_SPACE_RE.sub(" ", text)
    current = fold_extended_homoglyphs(current)
    if unicode_strict:
        try:
            from .confusables import normalize_confusables

            current = normalize_confusables(current)
        except Exception:
            pass
    current = strip_combining_marks(current)
    current = unicodedata.normalize("NFKC", current)
    current = ANSI_ESCAPE_RE.sub("", current)
    current = EMOJI_RE.sub(" ", current)
    return collapse_control_whitespace(current)


ROT13_VARIANT_PATTERN_IDS = frozenset(
    {"rot13-obfuscation", "ignore-instructions", "leetspeak-injection"}
)


def injection_match_variants(preprocessed: str, *, include_rot13: bool = False) -> list[str]:
    variants = {preprocessed, deleetspeak(preprocessed)}
    if include_rot13:
        r13 = rot13(preprocessed)
        variants.add(r13)
        variants.add(deleetspeak(r13))
    return list(variants)

File: src/utils/injection-preprocess.ts
/**
 * Text preprocessing for prompt-injection and semantic regex matching.
 * Closes unicode, boundary, ROT13, leetspeak, ANSI, and emoji evasion gaps.
 */
import { foldHomoglyphs, normalizeConfusables } from './confusables.js';

/** Zero-width / bidi — replaced with ASCII space so word boundaries survive. */
const INVISIBLE_TO_SPACE_RE =
  /[\u200B-\u200F\uFEFF\u00AD\u2060-\u2064\u061C\u180E\u034F\u17B4\u17B5\u202A-\u202E\u2800\uFE00-\uFE0F]/g;

/** Braille pattern blank and other unicode space separators. */
const UNICODE_SPACE_RE =
  /[\u00A0\u1680\u2000-\u200A\u202F\u205F\u3000\u3164]/g;

/** ANSI CSI and related escape sequences. */
const ANSI_ESCAPE_RE = /\x1b\[[0-9;]*[A-Za-z]|\x1b\].*?\x07/g;

/** Emoji blocks used as instruction padding. */
const EMOJI_RE =
  /[\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\u{2300}-\u{23FF}]/gu;

const LEET_MAP: Record<string, string> = {
  '0': 'o',
  '1': 'i',
  '3': 'e',
  '4': 'a',
  '5': 's',
  '7': 't',
  '@': 'a',
  '$': 's',
};

/** Fast Cyrillic/Greek + math alphanumeric (U+1D400–U+1D7FF subset). */
const EXTRA_HOMOGLYPHS: Record<number, string> = {
  0x0438: 'i', // и
  0x0433: 'r', // г (rough)
  0x043d: 'n', // н
  0x0441: 'c', // с
  0x0442: 't', // т
  0x0440: 'p', // р
  0x0432: 'b', // в
  0x043b: 'l', // л
  0x043c: 'm', // м
  0x0434: 'd', // д
  0x0455: 's', // ѕ
  0x04cf: 'l', // ӏ
};

export function rot13(text: string): string {
  return text.replace(/[a-zA-Z]/g, (ch) => {
    const base = ch <= 'Z' ? 65 : 97;
    return String.fromCharCode(((ch.charCodeAt(0) - base + 13) % 26) + base);
  });
}

export function deleetspeak(text: string): string {
  return text.replace(/[013457@$]/g, (ch) => LEET_MAP[ch] ?? ch);
}

export function stripCombiningMarks(text: string): string {
  return text.normalize('NFD').replace(/\p{M}/gu, '').normalize('NFC');
}

export function foldExtendedHomoglyphs(input: string): string {
  let out = foldHomoglyphs(input);
  let result = '';
  for (const ch of out) {
    const code = ch.codePointAt(0)!;
    if (code >= 0x1d400 && code <= 0x1d7ff) {
      const offset = code - 0x1d400;
      const latin = offset % 26;
      const isUpper = Math.floor(offset / 26) % 2 === 1;
      result += String.fromCharCode((isUpper ? 65 : 97) + latin);
      continue;
    }
    result += EXTRA_HOMOGLYPHS[code] ?? ch;
  }
  return result;
}

/**
 * Collapse control chars and exotic whitespace to a single ASCII space.
 */
export function collapseControlWhitespace(text: string): string {
  return text
    .replace(INVISIBLE_TO_SPACE_RE, ' ')
    .replace(UNICODE_SPACE_RE, ' ')
    .replace(/[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]/g, ' ')
    .replace(/[\t\r\n\f\v]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Full normalization pipeline before regex / injection rules run.
 */
export function preprocessForInjectionMatch(
  input: string,
  unicodeStrict = true,
): string {
  let current = input;
  current = current.replace(INVISIBLE_TO_SPACE_RE, ' ');
  current = foldExtendedHomoglyphs(current);
  if (unicodeStrict) {
    current = normalizeConfusables(current);
  }
  current = stripCombiningMarks(current);
  current = current.normalize('NFKC');
  current = current.replace(ANSI_ESCAPE_RE, '');
  current = EMOJI_RE.test(current) ? current.replace(EMOJI_RE, ' ') : current;
  current = collapseControlWhitespace(current);
  return current;
}

/** Variants for injection regex (leetspeak always; ROT13 only when requested). */
export function injectionMatchVariants(
  preprocessed: string,
  options?: { includeRot13?: boolean },
): string[] {
  const variants = new Set<string>();
  variants.add(preprocessed);
  variants.add(deleetspeak(preprocessed));
  if (options?.includeRot13) {
    const r13 = rot13(preprocessed);
    variants.add(r13);
    variants.add(deleetspeak(r13));
  }
  return [...variants];
}

const ROT13_VARIANT_PATTERN_IDS = new Set([
  'rot13-obfuscation',
  'ignore-instructions',
  'leetspeak-injection',
]);

export function shouldTryRot13Variant(patternId: string): boolean {
  return ROT13_VARIANT_PATTERN_IDS.has(patternId);
}
